xds/internal/balancer/xdsrouting/matcher_header.go

```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsrouting
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/metadata"
+)
+
+func mdValuesFromOutgoingCtx(ctx context.Context, key string) ([]string, bool) {
+	if ctx == nil {
+		return nil, false
+	}
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		return nil, false
+	}
+	vs, ok := md[key]
+	return vs, ok
+}
+
+type headerExactMatcher struct {
+	key   string
+	exact string
+}
+
+func newHeaderExactMatcher(key, exact string) *headerExactMatcher {
+	return &headerExactMatcher{key: key, exact: exact}
+}
+
+func (hem *headerExactMatcher) match(info balancer.PickInfo) bool {
+	vs, ok := mdValuesFromOutgoingCtx(info.Ctx, hem.key)
+	if !ok {
+		return false
+	}
+	for _, v := range vs {
+		if v == hem.exact {
+			return true
+		}
+	}
+	return false
+}
+
+func (hem *headerExactMatcher) Equal(m matcher) bool {
+	mm, ok := m.(*headerExactMatcher)
+	if !ok {
+		return false
+	}
+	return hem.key == mm.key && hem.exact == mm.exact
+}
+
+func (hem *headerExactMatcher) String() string {
+	return fmt.Sprintf("headerExact:%v:%v", hem.key, hem.exact)
+}
+
+type headerRegexMatcher struct {
+	key string
+	s   string
+	re  *regexp.Regexp
+}
+
+func newHeaderRegexMatcher(key, regexStr string) *headerRegexMatcher {
+	return &headerRegexMatcher{key: key, s: regexStr, re: regexp.MustCompile(regexStr)}
+}
+
+func (hrm *headerRegexMatcher) match(info balancer.PickInfo) bool {
```

Comment:
How about taking a `metadata.MD` instead of `balancer.PickInfo` for all header matchers. None of the other fields are used anyways. Also, it would make sense at the LB policy to extract the metadata from the PickInfo once, and pass the same thing to all header matchers instead of each of the header matchers repeating the same task.

Reply:
Actually, might also be worth using `FromOutgoingContextRaw`, if we can make that work without making it too ugly. Because this code path will be executed for every RPC.

Reply:
Done, after splitting headerMatcherInterface and pathMatcherInterface.